src/Features/Core/Portable/PdbSourceDocument/PdbSourceDocumentLoaderService.cs

```diff
@@ -19,86 +19,149 @@ namespace Microsoft.CodeAnalysis.PdbSourceDocument
     [Export(typeof(IPdbSourceDocumentLoaderService)), Shared]
     internal sealed class PdbSourceDocumentLoaderService : IPdbSourceDocumentLoaderService
     {
+        private const int SourceLinkTimeout = 1000;
+        private readonly ISourceLinkService? _sourceLinkService;
+
         [ImportingConstructor]
-        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-        public PdbSourceDocumentLoaderService()
+        [SuppressMessage("RoslynDiagnosticsReliability", "RS0033:Importing constructor should be [Obsolete]", Justification = "Used in test code")]
+        public PdbSourceDocumentLoaderService([Import(AllowDefault = true)] ISourceLinkService? sourceLinkService)
         {
+            _sourceLinkService = sourceLinkService;
         }
 
-        public Task<TextLoader?> LoadSourceDocumentAsync(SourceDocument sourceDocument, Encoding? defaultEncoding, CancellationToken cancellationToken)
+        public async Task<SourceFileInfo?> LoadSourceDocumentAsync(string tempFilePath, SourceDocument sourceDocument, Encoding encoding, IPdbSourceDocumentLogger? logger, CancellationToken cancellationToken)
         {
             // First we try getting "local" files, either from embedded source or a local file on disk
-            var stream = TryGetEmbeddedSourceStream(sourceDocument) ??
-                TryGetFileStream(sourceDocument);
+            // and if they don't work we call the debugger to download a file from SourceLink info
+            return TryGetEmbeddedSourceFile(tempFilePath, sourceDocument, encoding) ??
+                TryGetOriginalFile(sourceDocument, encoding) ??
+                await TryGetSourceLinkFileAsync(sourceDocument, encoding, logger, cancellationToken).ConfigureAwait(false);
+        }
+
+        private static SourceFileInfo? TryGetEmbeddedSourceFile(string tempFilePath, SourceDocument sourceDocument, Encoding encoding)
+        {
+            if (sourceDocument.EmbeddedTextBytes is null)
+                return null;
+
+            var filePath = Path.Combine(tempFilePath, Path.GetFileName(sourceDocument.FilePath));
+
+            // We might have already navigated to this file before, so it might exist, but
+            // we still need to re-validate the checksum and make sure its not the wrong file
+            if (File.Exists(filePath) &&
```

Comment:
interesting, is there really much gain from keeping the file on disk instead of pulling it out fresh every time? Seems like unnecessary complexity

Reply:
Probably not, but I don't think we save much by adding the complexity to figure out when to delete it either.

Reply:
I think I realized later that we are required to put it on disk for VS (as a comment later in the file)